client/service/src/client/wasm_overwrite.rs

```diff
@@ -0,0 +1,144 @@
+use sp_core::traits::RuntimeCode;
+use hash_db::Hasher;
+
+#[derive(Clone, Debug)]
+struct WasmBlob {
+	code: Vec<u8>,
+}
+
+impl WasmBlob {
+	fn new(code: Vec<u8>) -> Self {
+		Self { code }
+	}
+
+	fn runtime_code(&self, heap_pages: Option<u64>) -> RuntimeCode {
+		RuntimeCode {
+			code_fetcher: self,
+			hash: make_hash(self.code.as_slice()).encode(),
+			heap_pages,
+		}
+	}
+}
+
+/// Make a hash out of a byte string using the default rust hasher
+pub fn make_hash<K: std::hash::Hash + ?Sized>(val: &K) -> Vec<u8> {
+	let mut state = DefaultHasher::new();
+	val.hash(&mut state);
+	state.finish().to_le_bytes().to_vec()
+}
+
+impl FetchRuntimeCode for WasmBlob {
+	fn fetch_runtime_code<'a>(&'a self) -> Option<std::borrow::Cow<'a, [u8]>> {
+		Some(self.code.as_slice().into())
+	}
+}
+
+#[derive(Clone, Debug)]
+pub struct WasmOverwrite<E> {
+	// Map of runtime spec version -> Wasm Blob
+	overwrites: HashMap<u32, WasmBlob>,
+	executor: E,
+	enabled: bool,
+}
+
+impl<E> WasmOverwrite<E> 
+where 
+	E: RuntimeInfo + Clone + 'static 
+{
+	pub fn new<P>(path: P, enabled: bool, executor: E) -> Result<Self> 
+	where
+		P: AsRef<Path>,
+	{
+		let overwrites = Self::scrape_overwrites(path.as_ref(), &executor)?;
+		Ok(Self { overwrites, executor, enabled })
+	}
+	
+	/// Tries to replace the given `code` with an overwrite, if it exists.
+	/// If the overwrite does not exist, or overwrites are not enabled,
+	/// this function returns the original runtime code.
+	pub fn try_replace<'a, 'b: 'a, B, H>(
+		&'b self, 
+		code: RuntimeCode<'a>, 
+		state: &'a B,
+	) -> Result<RuntimeCode<'a>>
+	where
+		B: backend::Backend<H>,
+		H: Hasher
+	{
+		if !self.enabled {
+			return Ok(code);
+		}
+
+		let backend_code = code.fetch_runtime_code()
+			.ok_or(sp_blockchain::Error::Msg(format!("Runtime code could not be found in the backend")))?;
+		let heap_pages = state.storage(sp_core::storage::well_known_keys::HEAP_PAGES)
+			.ok()
+			.flatten()
+			.and_then(|d| Decode::decode(&mut &d[..]).ok()); 
+		let version = Self::runtime_version(&self.executor, &WasmBlob::new(backend_code.to_vec()), heap_pages)?;
+		
+		if let Some(runtime_code) = self.overwrites
+			.get(&version.spec_version)
+			.map(|w| w.runtime_code(heap_pages)) 
+		{
+			Ok(runtime_code)
+		} else {
+			Ok(code)
+		}
+	}
+
+	/// Scrapes a folder for WASM runtimes.
+    /// Returns a hashmap of the runtime version and wasm runtime code.
+	fn scrape_overwrites(dir: &Path, executor: &E) -> Result<HashMap<u32, WasmBlob>> {
+		let handle_err = |e: std::io::Error | -> sp_blockchain::Error {
+			sp_blockchain::Error::Msg(format!("{}", e.to_string()))
+		};
+		
+		let mut overwrites = HashMap::new(); 
+		if dir.is_dir() {
+			for entry in fs::read_dir(dir).map_err(handle_err)? {
+				let entry = entry.map_err(handle_err)?;
+				let path = entry.path();
+				let wasm = WasmBlob::new(fs::read(path).map_err(handle_err)?);
+				let version = Self::runtime_version(executor, &wasm, Some(128))?;
```

Comment:
HEAP_PAGES is a synomym for wasm linear memory pages and therefore there is no such thing as "native" `HEAP_PAGES` - the native runtime uses the same memory/allocator as the rest of the node.

Now, since this is a call to a particular function, `runtime_version` it might be fine. After all, this function in standard runtimes (standard = e.g. FRAME based) would just put some data through SCALE encode.

Ideally though, the version should be embedded in the wasm file as a section so that there won't be need in any code execution.

Reply:
> Ideally though, the version should be embedded in the wasm file as a section so that there won't be need in any code execution.

This would simplify the code for this PR quite a bit; do you know if there's a ticket for that?

Reply:
Nope, sorry, it was folklore only.

I've just created it https://github.com/paritytech/substrate/issues/7327